File: dplyr_tpc/h20.R
tpc_h20 <- function(input_func, collect_func = dplyr::collect) {
  supplier_ca <- input_func("supplier") %>%
    inner_join(
      input_func("nation") %>% filter(n_name == "CANADA"),
      by = c("s_nationkey" = "n_nationkey")
    ) %>%
    select(s_suppkey, s_name, s_address)

  part_forest <- input_func("part") %>%
    filter(grepl("^forest", p_name))

  partsupp_forest_ca <- input_func("partsupp") %>%
    semi_join(supplier_ca, c("ps_suppkey" = "s_suppkey")) %>%
    semi_join(part_forest, by = c("ps_partkey" = "p_partkey"))

  qty_threshold <- input_func("lineitem") %>%
    filter(
      l_shipdate >= as.Date("1994-01-01"),
      l_shipdate < as.Date("1995-01-01")
    ) %>%
    semi_join(partsupp_forest_ca, by = c("l_partkey" = "ps_partkey", "l_suppkey" = "ps_suppkey")) %>%
    group_by(l_suppkey) %>%
    summarise(qty_threshold = 0.5 * sum(l_quantity))

  partsupp_forest_ca_filtered <- partsupp_forest_ca %>%
    inner_join(
      qty_threshold,
      by = c("ps_suppkey" = "l_suppkey")
    ) %>%
    filter(ps_availqty > qty_threshold)

  supplier_ca %>%
    semi_join(partsupp_forest_ca_filtered, by = c("s_suppkey" = "ps_suppkey")) %>%
    select(s_name, s_address) %>%
    arrange(s_name) %>%
    collect_func()
}

File: ibis_tpc/h20.py
from .utils import add_date


def tpc_h20(con, COLOR='forest', DATE='1994-01-01', NATION='CANADA'):
    '''Potential Part Promotion Query (Q20)

    The Potential Part Promotion Query identifies suppliers in a particular
    nation having selected parts that may be candidates for a promotional
    offer.'''

    supplier = con.table('supplier')
    nation = con.table('nation')
    partsupp = con.table('partsupp')
    part = con.table('part')
    lineitem = con.table('lineitem')

    q1 = supplier.join(nation, supplier.S_NATIONKEY == nation.N_NATIONKEY)
    q1 = q1.materialize()

    q3 = part.filter([part.P_NAME.like(f'{COLOR}%')])
    q2 = partsupp

    q4 = lineitem.filter([
        lineitem.L_PARTKEY == q2.PS_PARTKEY,
        lineitem.L_SUPPKEY == q2.PS_SUPPKEY,
        lineitem.L_SHIPDATE >= DATE,
        lineitem.L_SHIPDATE < add_date(DATE, dy=1),
    ])

    q2 = q2.filter([
        partsupp.PS_PARTKEY.isin(q3.P_PARTKEY),
        partsupp.PS_AVAILQTY > .5*q4.L_QUANTITY.sum()
    ])

    q1 = q1.filter([
        q1.N_NAME == NATION,
        q1.S_SUPPKEY.isin(q2.PS_SUPPKEY)
    ])

    q1 = q1[q1.S_NAME, q1.S_ADDRESS]

    return q1.sort_by(q1.S_NAME)
